src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,425 @@
+                    _waitHandles[_numUserWaits + 1] = handle.Handle;
+                    _numUserWaits++;
+
+                    handle.WaitThread = this;
+                }
+                finally
+                {
+                    _registeredHandlesLock.Release();
+                }
+
+                _changeHandlesEvent.Set();
+                return true;
+            }
+
+            /// <summary>
+            /// Starts the wait thread.
+            /// </summary>
+            public void Start()
+            {
+                _waitHandles[0] = _changeHandlesEvent;
+                RuntimeThread waitThread = RuntimeThread.Create(WaitThreadStart);
+                waitThread.IsBackground = true;
+                waitThread.Start();
+            }
+
+            /// <summary>
+            /// Queues on the thread pool or executes directly a call to <see cref="UnregisterWait(object)"/>.
+            /// </summary>
+            /// <param name="handle">The handle to unregister.</param>
+            /// <remarks>
+            /// As per CoreCLR's behavior, if the user passes in an invalid <see cref="WaitHandle"/>
+            /// into <see cref="RegisteredWait.Unregister(WaitHandle)"/>, then the unregistration of the wait handle is blocking.
+            /// Otherwise, the unregistration of the wait handle is queued on the thread pool.
+            /// </remarks>
+            public void QueueOrExecuteUnregisterWait(RegisteredWait handle)
+            {
+                if (handle.IsBlocking)
+                {
+                    UnregisterWait(handle);
+                }
+                else
+                {
+                    ThreadPool.QueueUserWorkItem(UnregisterWait, handle);
+                }
+            }
+
+            /// <summary>
+            /// Unregister a wait handle.
+            /// </summary>
+            /// <param name="state">The wait handle to unregister.</param>
+            private void UnregisterWait(object state)
+            {
+                RegisteredWait handle = (RegisteredWait)state;
+
+                // TODO: Optimization: Try to unregister wait directly if it isn't being waited on.
+                _removesLock.Acquire();
+                try
+                {
+                    Debug.Assert(Array.IndexOf(_registeredWaits, handle) != -1 || handle.IsUnregistered);
+                    // If this handle is not already pending removal and hasn't already been removed
+                    if (Array.IndexOf(_pendingRemoves, handle) == -1 && !handle.IsUnregistered)
+                    {
+                        _pendingRemoves[_numPendingRemoves++] = handle;
+                        _changeHandlesEvent.Set(); // Tell the wait thread that there are changes pending.
+                    }
+                }
+                finally
+                {
+                    _removesLock.Release();
+                }
+
+                handle.SignalUserWaitHandle();
```

Comment:
In the sense we decided it doesn't have to wait for the WaitAny to complete in order to return from Unregister since I was worried about returning and allowing a wait handle to be disposed while it's being waited upon by WaitAny. But as I mentioned last time, if the wait handle being unregistered is also signaled and the signal wins the race (WaitAny wakes up due to a signal), it has to call the callback before the wait can be unregistered. For instance AutoResetEvent will auto-reset once its signal is observed by a thread, but that signal can't go lost and the callback is the only way to have it be observed by the user. So if the signal wins the race, the callback is going to be called and Unregister would have to wait for it before it can signal the provided event.